File: frontend/app/types/event.ts
/* frontend/app/types/event.ts */

import { ProposalOfficeAssignments } from './proposal';

export type EventUiStatus = 'LIVE' | 'PENDING' | 'COMPLETED' | 'UPCOMING' | 'CANCELLED' | 'ARCHIVED';
export type EventTaskPriority = 'low' | 'medium' | 'high';
export type EventTaskStatus = 'open' | 'in_progress' | 'done' | 'cancelled';
export type EventAnnouncementChannel = 'in_app';
export type EventIncidentSeverity = 'low' | 'medium' | 'high' | 'critical';
export type EventIncidentStatus = 'open' | 'in_review' | 'resolved' | 'closed';
export type FinalReportVisibility = 'private' | 'sponsors' | 'government' | 'public';

export interface PastEvent {
  id: string;
  title: string;
  date: string;
  event_type?: string;
}

export interface EventStats {
  totalEvents: number;
  thisMonth: number;
  drafts: number;
}

export interface EventProgress {
  proposal: number;
  approval: number;
  vendors: number;
  booking: number;
}

export interface EventListItem {
  id: string;
  proposal_id: string;
  title: string;
  event_type?: string;
  location: string;
  date: string;
  start_date?: string;
  end_date?: string;
  status: EventUiStatus;
  backend_status: string;
  image?: string;
  progress?: EventProgress;
  permit_number?: string | null;
  booking_required?: boolean;
  booking_status?: string;
  booked_count?: number;
  remaining_slots?: number;
}

export interface ApprovedProposal {
  id: string;
  event_id?: string | null;
  title: string;
  location?: string;
  status: string;
  approved_date: string;
  created_at: string;
  event_type?: string;
}

export interface EventBudgetItem {
  id?: string | null;
  name: string;
  estimated_cost: number;
  actual_cost?: number | null;
  notes?: string | null;
}

export interface EventRecord {
  id: string;
  organizer_id: string;
  proposal_id: string;
  permit_id?: string | null;
  permit_number?: string | null;
  title: string;
  description?: string | null;
  category?: string | null;
  location?: string | null;
  capacity?: number | null;
  start_date?: string | null;
  end_date?: string | null;
  visibility: string;
  booking_required: boolean;
  vip_list: string[];
  program_schedule_summary?: string | null;
  requires_permit: boolean;
  status: string;
  venue_status: string;
  booking_status: string;
  booking_opens_at?: string | null;
  booking_closes_at?: string | null;
  allow_waitlist: boolean;
  required_attendee_fields: string[];
  booked_count: number;
  remaining_slots: number;
  survey_status: string;
  final_report_status: string;
  budget_currency: string;
  budget_items: EventBudgetItem[];
  budget_total_estimated: number;
  office_assignments?: ProposalOfficeAssignments | null;
  published_at?: string | null;
  live_started_at?: string | null;
  completed_at?: string | null;
  archived_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface EventCreateFromProposalResponse {
  event_id: string;
  proposal_id: string;
  status: string;
  message: string;
}

export interface EventBudgetUpdatePayload {
  items: EventBudgetItem[];
  currency: string;
}

export interface EventScheduleItemRecord {
  id: string;
  event_id: string;
  session_title: string;
  description?: string | null;
  start_time: string;
  end_time: string;
  speaker_id?: string | null;
  room_location?: string | null;
  is_ai_suggestion: boolean;
  created_at: string;
  updated_at: string;
}

export interface EventScheduleAiDraftPayload {
  event_type: string;
  duration_days: number;
  start_time: string;
}

export interface AiScheduleDraftItem {
  title: string;
  start_time: string;
  end_time: string;
  category?: string | null;
  description?: string | null;
  order_index: number;
  is_ai_suggestion: boolean;
  applied: boolean;
}

export interface AiScheduleDraftResponse {
  id: string;
  event_id: string;
  organizer_id: string;
  input_constraints: Record<string, any>;
  provider: string;
  status: string;
  generated_items: AiScheduleDraftItem[];
  created_at: string;
  expires_at: string;
  can_apply: boolean;
}

export interface EventScheduleCreatePayload {
  session_title: string;
  description?: string;
  start_time: string;
  end_time: string;
  speaker_id?: string;
  room_location?: string;
  is_ai_suggestion?: boolean;
}

export interface EventScheduleUpdatePayload {
  session_title?: string;
  description?: string;
  start_time?: string;
  end_time?: string;
  speaker_id?: string;
  room_location?: string;
  is_ai_suggestion?: boolean;
}

export interface TicketTypeRecord {
  id: string;
  event_id: string;
  name: string;
  description?: string | null;
  price: number;
  quantity: number;
  sold_quantity: number;
  reserved_quantity: number;
  remaining_quantity: number;
  currency: string;
  sales_start?: string | null;
  sales_end?: string | null;
  seat_mode: 'general' | 'assigned';
  visibility: 'public' | 'private';
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface TicketTypeCreatePayload {
  name: string;
  description?: string;
  price: number;
  quantity: number;
  currency?: string;
  sales_start?: string;
  sales_end?: string;
  seat_mode?: 'general' | 'assigned';
  visibility?: 'public' | 'private';
  is_active?: boolean;
}

export interface TicketTypeUpdatePayload {
  name?: string;
  description?: string;
  price?: number;
  quantity?: number;
  currency?: string;
  sales_start?: string;
  sales_end?: string;
  seat_mode?: 'general' | 'assigned';
  visibility?: 'public' | 'private';
  is_active?: boolean;
}

export interface TicketCheckoutRequest {
  ticket_type_id: string;
  quantity: number;
  attendee_name?: string;
  attendee_email?: string;
  notes?: string;
  attendee_profile?: Record<string, string>;
}

export interface TicketPaymentConfirmRequest {
  payment_reference_id?: string;
  payment_method?: string;
}

// ─── Phase 2: Venue Search & Reservation ─────────────────────────────────────

export type VenueReservationStatus =
  | 'requested'
  | 'provider_accepted'
  | 'offered_alternative'
  | 'organizer_confirmed'
  | 'confirmed'
  | 'declined'
  | 'cancelled';

export type VenueReservationPaymentStatus =
  | 'not_required'
  | 'deposit_pending'
  | 'deposit_funded'
  | 'satisfied';

export interface VenueListingSearchResponse {
  id: string;
  venue_name: string;
  city: string;
  location: string | null;
  capacity: number;
  estimated_cost: number | null;
  deposit_amount: number | null;
  currency: string;
  available: boolean;
  is_reservable: boolean;
  description: string | null;
  notes: string | null;
  vendor: { vendor_id: string; business_name: string | null } | null;
}

export interface VenueSearchResult {
  event_id: string;
  date_from?: string | null;
  date_to?: string | null;
  city?: string | null;
  venues: VenueListingSearchResponse[];
}

/** Phase 2: creation now uses venue_listing_id instead of free-text */
export interface VenueReservationCreatePayload {
  vendor_id: string;
  requested_start: string;
  requested_end: string;
  estimated_cost?: number;
  notes?: string;
}

export interface VenueReservationOrganizerConfirmPayload {
  confirmation_notes?: string;
}

export interface VenueReservationCancelPayload {
  cancellation_notes?: string;
}

export interface VenueReservationDepositUpdatePayload {
  payment_milestone_status: VenueReservationPaymentStatus;
  payment_reference_id?: string;
  notes?: string;
}

/** Phase 2 full handshake reservation record */
export interface VenueReservationRecord {
  id: string;
  event_id: string;
  vendor_id: string;
  vendor_user_id: string;
  venue_name: string;
  city: string;
  location: string | null;
  requested_start: string;
  requested_end: string;
  requested_capacity: number | null;
  estimated_cost: number | null;
  deposit_amount: number | null;
  currency: string;
  notes: string | null;
  provider_action: string | null;
  provider_response_notes: string | null;
  failure_reason: string | null;
  organizer_confirmation_notes: string | null;
  cancellation_notes: string | null;
  proposed_start: string | null;
  proposed_end: string | null;
  proposed_cost: number | null;
  proposed_deposit_amount: number | null;
  agreed_start: string | null;
  agreed_end: string | null;
  agreed_cost: number | null;
  agreed_deposit_amount: number | null;
  status: VenueReservationStatus;
  payment_milestone_status: VenueReservationPaymentStatus;
  payment_reference_id: string | null;
  deposit_funded_at: string | null;
  deposit_satisfied_at: string | null;
  alternative_suggestions: VenueListingSearchResponse[];
  alternative_suggestions_generated_at: string | null;
  created_at: string;
  updated_at: string;
  failed_at: string | null;
  provider_responded_at: string | null;
  organizer_confirmed_at: string | null;
  confirmed_at: string | null;
  cancelled_at: string | null;
}

export interface EventTeamInvitationCreatePayload {
  email: string;
  assigned_role: string;
  display_name?: string;
}

export interface EventTeamInvitationRecord {
  id: string;
  event_id: string;
  email: string;
  assigned_role: string;
  display_name?: string | null;
  invited_by_user_id: string;
  status: string;
  token: string;
  created_at: string;
  updated_at: string;
  accepted_at?: string | null;
}

export interface EventTeamMemberRecord {
  id: string;
  event_id: string;
  user_id?: string | null;
  email: string;
  full_name?: string | null;
  assigned_role: string;
  status: string;
  joined_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface EventTaskCreatePayload {
  title: string;
  description?: string;
  assignee_user_id?: string;
  assignee_email?: string;
  due_date?: string;
  priority?: EventTaskPriority;
}

export interface EventTaskUpdatePayload {
  title?: string;
  description?: string;
  assignee_user_id?: string;
  assignee_email?: string;
  due_date?: string;
  priority?: EventTaskPriority;
  status?: EventTaskStatus;
}

export interface EventTaskRecord {
  id: string;
  event_id: string;
  title: string;
  description?: string | null;
  assignee_user_id?: string | null;
  assignee_email?: string | null;
  due_date?: string | null;
  priority: EventTaskPriority;
  status: EventTaskStatus;
  created_at: string;
  updated_at: string;
}

export interface BookingSettingsUpdatePayload {
  booking_required: boolean;
  booking_opens_at?: string | null;
  booking_closes_at?: string | null;
  allow_waitlist: boolean;
  required_attendee_fields: string[];
}

export interface EventBookingCreatePayload {
  slots_requested: number;
  attendee_name?: string;
  attendee_email?: string;
  notes?: string;
  attendee_profile?: Record<string, string>;
}

export interface EventBookingRecord {
  id: string;
  event_id: string;
  booking_reference: string;
  event_title?: string | null;
  event_location?: string | null;
  event_start_date?: string | null;
  event_end_date?: string | null;
  attendee_id: string;
  attendee_name?: string | null;
  attendee_email?: string | null;
  slots_requested: number;
  notes?: string | null;
  attendee_profile: Record<string, string>;
  qr_code?: string | null;
  qr_code_image_url?: string | null;
  check_in_pass_image_url?: string | null;
  booking_status: string;
  check_in_status: string;
  checked_in_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface BadgeGeneratePayload {
  booking_ids?: string[] | null;
  include_unchecked_in?: boolean;
}

export interface BadgeRecord {
  id: string;
  event_id: string;
  booking_id: string;
  attendee_id: string;
  attendee_name?: string | null;
  attendee_email?: string | null;
  badge_code: string;
  role_label: string;
  generated_at: string;
}

export interface CheckInScanPayload {
  qr_code: string;
}

export interface AnnouncementCreatePayload {
  audience_segment: 'confirmed_bookings';
  subject: string;
  body: string;
  channel?: EventAnnouncementChannel;
  send_at?: string | null;
}

export interface AnnouncementRecord {
  id: string;
  event_id: string;
  audience_segment: string;
  subject: string;
  body: string;
  channel: EventAnnouncementChannel | string;
  send_at?: string | null;
  status: string;
  created_at: string;
  updated_at: string;
  sent_at?: string | null;
  recipient_count: number;
  delivered_count: number;
  delivery_warning?: string | null;
}

export interface AnnouncementDeliveryRecord {
  id: string;
  announcement_id: string;
  event_id: string;
  recipient_user_id: string;
  recipient_name?: string | null;
  recipient_email?: string | null;
  booking_id?: string | null;
  subject: string;
  body: string;
  status: string;
  delivered_at?: string | null;
  read_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface ManualAnnouncementRunResponse {
  announcement: AnnouncementRecord;
  deliveries: AnnouncementDeliveryRecord[];
}

export interface IncidentCreatePayload {
  type: string;
  severity?: EventIncidentSeverity;
  time?: string | null;
  description: string;
  photos?: string[];
}

export interface IncidentUpdatePayload {
  severity?: EventIncidentSeverity;
  description?: string;
  photos?: string[];
  escalation_status?: string;
  status?: EventIncidentStatus;
}

export interface IncidentRecord {
  id: string;
  event_id: string;
  type: string;
  severity: EventIncidentSeverity | string;
  time: string;
  description: string;
  photos: string[];
  created_by_user_id: string;
  escalation_status: string;
  status: EventIncidentStatus | string;
  created_at: string;
  updated_at: string;
  resolved_at?: string | null;
}

export interface FeedbackSendPayload {
  audience_segment?: string;
  scheduled_for?: string | null;
  custom_questions?: string[];
}

export interface FeedbackSummaryRecord {
  event_id: string;
  survey_status: string;
  response_count: number;
  average_rating?: number | null;
  average_vendor_rating?: number | null;
  average_nps?: number | null;
}

export interface FinalReportCreatePayload {
  timeline_summary: string;
  total_costs?: number | null;
  vendors_used?: string[];
  lessons_learned: string;
  visibility?: FinalReportVisibility;
  benchmark_notes?: string;
}

export interface FinalReportUpdatePayload {
  timeline_summary?: string;
  total_costs?: number | null;
  vendors_used?: string[];
  lessons_learned?: string;
  visibility?: FinalReportVisibility;
  benchmark_notes?: string;
  status?: string;
}

export interface FinalReportRecord {
  id: string;
  event_id: string;
  timeline_summary: string;
  total_costs?: number | null;
  vendors_used: string[];
  lessons_learned: string;
  visibility: FinalReportVisibility | string;
  benchmark_notes?: string | null;
  status: string;
  created_at: string;
  updated_at: string;
  published_at?: string | null;
}

export interface EventUpdatePayload {
  description?: string;
  visibility?: 'public' | 'private';
  booking_required?: boolean;
  vip_list?: string[];
  program_schedule_summary?: string;
}

export const EVENT_STATUS_CONFIG: Record<EventUiStatus, { label: string; bgClass: string }> = {
  LIVE: { label: 'LIVE', bgClass: 'bg-green-500 text-white' },
  PENDING: { label: 'PENDING', bgClass: 'bg-amber-500 text-white' },
  COMPLETED: { label: 'COMPLETED', bgClass: 'bg-slate-500 text-white' },
  UPCOMING: { label: 'UPCOMING', bgClass: 'bg-blue-500 text-white' },
  CANCELLED: { label: 'CANCELLED', bgClass: 'bg-red-500 text-white' },
  ARCHIVED: { label: 'ARCHIVED', bgClass: 'bg-zinc-700 text-white' },
};

File: backend/app/schemas/event.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, ConfigDict, model_validator

from app.models.event_states import (
    BookingStatus,
    EventStatus,
    FinalReportStatus,
    SurveyStatus,
)
from app.schemas.proposal import ProposalOfficeAssignments


class EventBudgetItem(BaseModel):
    id: str | None = None
    name: str
    actual_cost: float | None = Field(default=None, ge=0)
    notes: str | None = None


class EventBudgetUpdate(BaseModel):
    budget_amount: float = Field(..., ge=0)
    items: list[EventBudgetItem] = []
    currency: str = "ETB"


class EventScheduleItemCreate(BaseModel):
    session_title: str
    description: str | None = None
    start_time: datetime
    end_time: datetime
    speaker_id: str | None = None
    room_location: str | None = None
    is_ai_suggestion: bool = False


class EventScheduleItemUpdate(BaseModel):
    session_title: str | None = None
    description: str | None = None
    start_time: datetime | None = None
    end_time: datetime | None = None
    speaker_id: str | None = None
    room_location: str | None = None
    is_ai_suggestion: bool | None = None


class EventScheduleItemResponse(BaseModel):
    id: str
    event_id: str
    session_title: str
    description: str | None = None
    start_time: datetime | None = None
    end_time: datetime | None = None
    speaker_id: str | None = None
    room_location: str | None = None
    is_ai_suggestion: bool = False
    created_at: datetime
    updated_at: datetime


class EventAIDraftRequest(BaseModel):
    duration_days: int | None = Field(default=None, ge=1, le=30)
    start_time: str | None = None
    sessions_per_day: int = Field(default=4, ge=1, le=12)


# Ticket types and ticket checkout endpoints were removed — bookings use the
# existing booking endpoints and `ticket_purchase_collection` for confirmed
# attendee records.


class EventTeamInvitationCreate(BaseModel):
    email: str
    assigned_role: str
    display_name: str | None = None


class EventTeamInvitationResponse(BaseModel):
    id: str
    event_id: str
    email: str
    assigned_role: str
    display_name: str | None = None
    invited_by_user_id: str
    status: str
    token: str
    created_at: datetime
    updated_at: datetime
    accepted_at: datetime | None = None


class EventTeamMemberResponse(BaseModel):
    id: str
    event_id: str
    user_id: str | None = None
    email: str
    full_name: str | None = None
    assigned_role: str
    status: str
    joined_at: datetime | None = None
    created_at: datetime
    updated_at: datetime

class EventTaskCreate(BaseModel):
    title: str
    description: str | None = None
    assignee_user_id: str | None = None
    assignee_email: str | None = None
    due_date: datetime | None = None
    priority: Literal["low", "medium", "high"] = "medium"
    payout_amount: float | None = Field(default=None, ge=0)


class EventTaskUpdate(BaseModel):
    title: str | None = None
    description: str | None = None
    assignee_user_id: str | None = None
    assignee_email: str | None = None
    due_date: datetime | None = None
    priority: Literal["low", "medium", "high"] | None = None
    status: Literal["open", "in_progress", "pending_approval", "done", "cancelled"] | None = None
    payout_amount: float | None = Field(default=None, ge=0)


class EventTaskResponse(BaseModel):
    id: str
    event_id: str
    title: str
    description: str | None = None
    assignee_user_id: str | None = None
    assignee_email: str | None = None
    due_date: datetime | None = None
    priority: str
    status: str
    payout_amount: float | None = None
    # Escrow fields
    escrow_locked: bool = False
    escrow_amount: float = 0.0
    escrow_locked_at: datetime | None = None
    # Workspace tracking
    workspace_open: bool = False
    workspace_opened_at: datetime | None = None
    workspace_closed: bool = False
    workspace_closed_at: datetime | None = None
    # Negotiation phase gate
    negotiation_phase_locked: bool = False
    negotiation_locked_at: datetime | None = None
    # Rejection flow
    rejection_note: str | None = None
    rejected_at: datetime | None = None
    rejection_count: int = 0
    created_at: datetime
    updated_at: datetime


class TaskRejectPayload(BaseModel):
    """Organizer must provide a mandatory note when rejecting a task."""
    note: str = Field(..., min_length=10, description="Mandatory explanation of what needs to be fixed.")


class BookingSettingsUpdate(BaseModel):
    visibility: Literal["public", "private"] | None = None
    booking_required: bool = True
    booking_opens_at: datetime | None = None
    booking_closes_at: datetime | None = None
    allow_waitlist: bool = False
    required_attendee_fields: list[str] = []

    @model_validator(mode="after")
    def validate_booking_window(self) -> "BookingSettingsUpdate":
        """booking_closes_at must be strictly after booking_opens_at when both are set."""
        opens  = self.booking_opens_at
        closes = self.booking_closes_at
        if opens and closes:
            # Strip timezone info for comparison (stored as naive UTC in MongoDB)
            opens_cmp  = opens.replace(tzinfo=None)  if opens.tzinfo  else opens
            closes_cmp = closes.replace(tzinfo=None) if closes.tzinfo else closes
            if closes_cmp <= opens_cmp:
                raise ValueError(
                    "booking_closes_at must be after booking_opens_at. "
                    f"Got opens={opens_cmp}, closes={closes_cmp}."
                )
        return self


class EventBookingCreate(BaseModel):
    slots_requested: int = Field(default=1, ge=1, le=20)
    attendee_name: str | None = None
    attendee_email: str | None = None
    notes: str | None = None
    attendee_profile: dict[str, str] = {}


class EventBookingResponse(BaseModel):
    id: str
    event_id: str
    booking_reference: str
    event_title: str | None = None
    event_location: str | None = None
    event_start_date: datetime | None = None
    event_end_date: datetime | None = None
    attendee_id: str
    attendee_name: str | None = None
    attendee_email: str | None = None
    slots_requested: int
    notes: str | None = None
    attendee_profile: dict[str, str] = {}
    qr_code: str | None = None
    qr_code_image_url: str | None = None
    check_in_pass_image_url: str | None = None
    booking_status: str
    check_in_status: str
    checked_in_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class CheckInScanPayload(BaseModel):
    qr_code: str


class BadgeGeneratePayload(BaseModel):
    booking_ids: list[str] | None = None
    include_unchecked_in: bool = True


class BadgeResponse(BaseModel):
    id: str
    event_id: str
    booking_id: str
    attendee_id: str
    attendee_name: str | None = None
    attendee_email: str | None = None
    badge_code: str
    role_label: str = "Attendee"
    generated_at: datetime


class AnnouncementCreate(BaseModel):
    audience_segment: str = "confirmed_bookings"
    subject: str
    body: str
    channel: str = "in_app"
    send_at: datetime | None = None


class AnnouncementResponse(BaseModel):
    id: str
    event_id: str
    audience_segment: str
    subject: str
    body: str
    channel: str
    send_at: datetime | None = None
    status: str
    created_at: datetime
    updated_at: datetime
    sent_at: datetime | None = None
    recipient_count: int = 0
    delivered_count: int = 0
    delivery_warning: str | None = None


class AnnouncementDeliveryResponse(BaseModel):
    id: str
    announcement_id: str
    event_id: str
    recipient_user_id: str
    recipient_name: str | None = None
    recipient_email: str | None = None
    booking_id: str | None = None
    subject: str
    body: str
    status: str
    delivered_at: datetime | None = None
    read_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class ManualAnnouncementRunResponse(BaseModel):
    announcement: AnnouncementResponse
    deliveries: list[AnnouncementDeliveryResponse]


class IncidentCreate(BaseModel):
    type: str
    severity: Literal["low", "medium", "high", "critical"] = "medium"
    time: datetime | None = None
    description: str
    photos: list[str] = []


class IncidentUpdate(BaseModel):
    severity: Literal["low", "medium", "high", "critical"] | None = None
    description: str | None = None
    photos: list[str] | None = None
    escalation_status: str | None = None
    status: Literal["open", "in_review", "resolved", "closed"] | None = None


class IncidentResponse(BaseModel):
    id: str
    event_id: str
    type: str
    severity: str
    time: datetime
    description: str
    photos: list[str] = []
    created_by_user_id: str
    escalation_status: str
    status: str
    created_at: datetime
    updated_at: datetime
    resolved_at: datetime | None = None


class FeedbackSendPayload(BaseModel):
    audience_segment: str = "all"
    scheduled_for: datetime | None = None
    custom_questions: list[str] = []


class FeedbackResponseCreate(BaseModel):
    rating: int = Field(..., ge=1, le=5)
    nps_score: int | None = Field(default=None, ge=0, le=10)
    comments: str | None = None
    vendor_rating: int | None = Field(default=None, ge=1, le=5)


class FeedbackResponseRecord(BaseModel):
    id: str
    event_id: str
    attendee_id: str
    rating: int
    nps_score: int | None = None
    comments: str | None = None
    vendor_rating: int | None = None
    created_at: datetime


class FeedbackSummaryResponse(BaseModel):
    event_id: str
    survey_status: SurveyStatus
    response_count: int
    average_rating: float | None = None
    average_vendor_rating: float | None = None
    average_nps: float | None = None


class FinalReportCreate(BaseModel):
    timeline_summary: str
    total_costs: float | None = Field(default=None, ge=0)
    vendors_used: list[str] = []
    lessons_learned: str
    visibility: Literal["private", "sponsors", "government", "public"] = "private"
    benchmark_notes: str | None = None


class FinalReportUpdate(BaseModel):
    timeline_summary: str | None = None
    total_costs: float | None = Field(default=None, ge=0)
    vendors_used: list[str] | None = None
    lessons_learned: str | None = None
    visibility: Literal["private", "sponsors", "government", "public"] | None = None
    benchmark_notes: str | None = None
    status: FinalReportStatus | None = None


class FinalReportResponse(BaseModel):
    id: str
    event_id: str
    timeline_summary: str
    total_costs: float | None = None
    vendors_used: list[str] = []
    lessons_learned: str
    visibility: str
    benchmark_notes: str | None = None
    status: FinalReportStatus
    created_at: datetime
    updated_at: datetime
    published_at: datetime | None = None


class EventCreateFromProposalResponse(BaseModel):
    event_id: str
    proposal_id: str
    status: EventStatus
    message: str


class EventUpdate(BaseModel):
    description: str | None = None
    visibility: Literal["public", "private"] | None = None
    booking_required: bool | None = None
    vip_list: list[str] | None = None
    program_schedule_summary: str | None = None


class EventResponse(BaseModel):
    id: str
    organizer_id: str
    proposal_id: str
    permit_id: str | None = None
    permit_number: str | None = None
    title: str
    description: str | None = None
    category: str | None = None
    location: str | None = None
    capacity: int | None = None
    start_date: datetime | None = None
    end_date: datetime | None = None
    visibility: str = "public"
    booking_required: bool = False
    vip_list: list[str] = []
    program_schedule_summary: str | None = None
    requires_permit: bool = True
    status: EventStatus
    venue_status: str
    booking_status: BookingStatus
    ticketing_status: str = "disabled"
    booking_opens_at: datetime | None = None
    booking_closes_at: datetime | None = None
    allow_waitlist: bool = False
    required_attendee_fields: list[str] = []
    booked_count: int = 0
    remaining_slots: int = 0
    survey_status: SurveyStatus
    final_report_status: FinalReportStatus
    budget_currency: str = "ETB"
    budget_items: list[EventBudgetItem] = []
    budget_amount: float = 0.0
    office_assignments: ProposalOfficeAssignments | None = None
    published_at: datetime | None = None
    live_started_at: datetime | None = None
    completed_at: datetime | None = None
    archived_at: datetime | None = None
    created_at: datetime
    updated_at: datetime
class VipReservationCreate(BaseModel):
    vip_name: str
    hotel_name: str
    vip_email: str | None = None
    notes: str | None = None

class VipReservationResponse(VipReservationCreate):
    id: str
    event_id: str
    created_at: datetime
    
    model_config = ConfigDict(from_attributes=True)


class EventCancelRequest(BaseModel):
    reason: str


class EventPostponeRequest(BaseModel):
    new_start_date: datetime
    new_end_date: datetime
    reason: str | None = None
